### DISCUSSION TITLE
Node version

### DISCUSSION THREAD

**Discussion by mzuniga51** [tagyoureit/nodejs-poolController#648] [General]
Hey.   I know 12 is the latest supported version of Node for njsPC.   Problem is some of my Homebridge plugins ask for version 16 or even higher.  When I update Node to the latest version njsPC stops working.  What is the last version of node we know njsPC works with?  Thanks

**Comment by tagyoureit** [comment]
I'm getting ready to push the 8.0.0 (both njsPC and dP) to the master branch but it needs some QA testing.  It should work with node 14, 16 and 18.  If you have a few mins, please do test it out.

For each app,
1. `git checkout 8.0.0`
2. `npm i`  
3. restart the app

**Reply by arrmo** [reply]
> I'm getting ready to push the 8.0.0 (both njsPC and dP) to the master branch but it needs some QA testing.

Thanks @tagyoureit! Is this ready for Virtual Circuit testing?

**Reply by tagyoureit** [reply]
Yes, check it out.  It has some fixes for the virtual circuits.

**Reply by arrmo** [reply]
> Yes, check it out. It has some fixes for the virtual circuits.

Will do! Out of town (EUR) right now, but when I get back this weekend 😄

**Comment by mzuniga51** [comment]
Will do.  Thanks!

**Comment by arrmo** [comment]
Hmmm ... OK, back, and realizing I need to update nodejs. NP, trying to install v16, but seeing this error - seen it before?
```
Unpacking nodejs (16.18.1-deb-1nodesource1) ...
dpkg: error processing archive /var/cache/apt/archives/nodejs_16.18.1-deb-1nodesource1_amd64.deb (--unpack):
 trying to overwrite '/usr/include/node/common.gypi', which is also in package libnode-dev 12.22.9~dfsg-1ubuntu3
dpkg-deb: error: paste subprocess was killed by signal (Broken pipe)
Errors were encountered while processing:
 /var/cache/apt/archives/nodejs_16.18.1-deb-1nodesource1_amd64.deb
E: Sub-process /usr/bin/dpkg returned an error code (1)
```
Thanks!

**Reply by arrmo** [reply]
Found it, make sure to `sudo apt remove libnode72` first 😆

**Comment by arrmo** [comment]
FYI, tried out the latest - turned on the spa, with my logic ... heater fired => virtual circuits, only the "Any heater" showed, up, not "Spa heater"?

Thanks!

**Comment by bmn001** [comment]
Brand new user here.

I saw a few people using this project on a Pi Zero W, so that's what I'm using.

I wasn't able to get 7.x installed due to a dependency error (I'm running Node 16.19.0 and npm 8.19.3) so I was glad to find this thread.

I'm not able to get this going.  I can clone the repo and run through the install with no errors.  But when I launch it, I get an out of memory issue after a few minutes:

```
pi@extra:~/nodejs-poolController $ npm start&
[1] 3298
pi@extra:~/nodejs-poolController $ 
> nodejs-poolcontroller@8.0.0 start
> npm run build && node dist/app.js


> nodejs-poolcontroller@8.0.0 build
> tsc


<--- Last few GCs --->

[3323:0x4d13da0]   608934 ms: Scavenge 123.4 (128.9) -> 123.0 (128.9) MB, 9.3 / 0.0 ms  (average mu = 0.186, current mu = 0.097) allocation failure 
[3323:0x4d13da0]   609106 ms: Scavenge 123.5 (128.9) -> 123.0 (128.9) MB, 6.0 / 0.0 ms  (average mu = 0.186, current mu = 0.097) allocation failure 
[3323:0x4d13da0]   613989 ms: Mark-sweep 123.7 (129.1) -> 122.6 (129.1) MB, 4631.8 / 0.0 ms  (average mu = 0.244, current mu = 0.298) allocation failure scavenge might not succeed


<--- JS stacktrace --->

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
Aborted
```

FWIW the 8.0.0 branch of dashPanel works fine.  But it would work a lot better with the poolController backend running ;)

On an unrelated note - dashPanel thinks my network router (an Eero Pro 6) is a nodejs-PoolController.

**Reply by tagyoureit** [reply]
Hi, yes 7.x should be installed with Node 12/14 and 8.x with Node 16/18.  You could try shutting down all applications and building njsPC separately with `npm run build`.  Then start the app with `node dist/app.js`.  The will prevent the compiling every time and if that's when it's running out of memory you might avoid the issue.  You can do the same with dashPanel and avoid compiling as well.  

If this doesn't work, you may be SOL with the Pi Zero.  Another user was able to get njsPC/dp running [thread](https://github.com/tagyoureit/nodejs-poolController/discussions/605) but had issues with REM and eventually gave up.

**Reply by Telemachus1977** [reply]
Actually I have REM, njsPC and Dashpanel all running well on a Pi Zero W for several weeks now, as long as the I2C bus is not running.  GPIO relays work fine.

You MUST build all three apps separately as described and then they can all be started together.

**Reply by bmn001** [reply]
Building it beforehand with nothing else running on the hardware crashes with the same error, unfortunately. 8.0.0 may just not be doable on a Pi Zero W. I've got a spare Pi 4 laying around that I may end up using, but was hoping to find a use for this inexpensive Zero.

@Telemachus1977 what versions of node/npm do you have running on your Zero W? I'll give that a shot before giving up.

**Reply by Telemachus1977** [reply]
I have a disc image with it fully working that I could share.  It's made for a 32GB SD card.

**Reply by bmn001** [reply]
No need to go to all that trouble but I appreciate the offer. I can work it out but having those versions would be helpful! :)

**Reply by rstrouse** [reply]
I am guessing the I2c interface pushes the limits of the Pi Zero.   If we ever see stock again I am going to the the Pi Zero 2 to see how far I can push it.  For those situations where an ESP32 just won't do.

**Reply by Telemachus1977** [reply]
I have no doubt the pi zero W 2 would run this just fine.

**Reply by bmn001** [reply]
I made some progress getting this to work. Still on Node 16, npm 8.

Running ```export NODE_OPTIONS=--max_old_space_size=256``` first will limit Node to using just half of the Pi Zero W's 512mb. After that, ```npm run build``` won't run out of memory!

But it will give a different error:

```pi@extra:~/nodejs-poolController $ npm run build

> nodejs-poolcontroller@8.0.0 build
> tsc

controller/comms/ScreenLogic.ts:86:39 - error TS2345: Argument of type 'string' is not assignable to parameter of type 'number'.

86         this._client.init(systemName, unit.ipAddr, unit.port, password);
                                         ~~~~~~~~~~~


Found 1 error in controller/comms/ScreenLogic.ts:86```

**Reply by bmn001** [reply]
Despite the error, it's running! While I do have a Screenlogic controller on my network, I don't have a Screenlogic tab under the Controller tab (and I'm still waiting for an RS-485 adapter to arrive) but I can click around in dashPanel and things seem pretty responsive. Progress!

**Comment by vj951** [comment]
Hello,
I am in the same boat. I had a pi zero W that worked perfectly last season, this season I kept getting error "JavaScript heap out of memory". I tried zero 2W, same issue. Then I tried max-old-space trick, and it worked. 

However, the max-old-space command does not survive the reboot. Therefore, any auto-restart with pm2 throws error as the current script does the full rebuild. 

My questions:
1. Is there anyway to include "export NODE_OPTIONS=--max_old_space_size=256" in pm2 script
2. Use "export NODE_OPTIONS=--max_old_space_size=256" at a system level permanently or somehow apply it at every reboot?
3. "npm run start:cached" command works for me, but I can't figure out how to call it from pm2 automatically. I can manually update/rebuild once in a while

I am only using what's in the wiki. I am not tech savvy, but will follow instructions to the dot.

Thanks

**Reply by tagyoureit** [reply]
1/2. You can either add it to your shell profile:
- For bash: ~/.bashrc or ~/.bash_profile
- For zsh: ~/.zshrc
or add it to the pm2 environment file (see Attributes available under https://pm2.keymetrics.io/docs/usage/application-declaration/).

3.  the start:cached script just runs `node dist/index` file.  You can either edit pm2 to run that script, or just call node directly with the file as the argument.  See https://github.com/tagyoureit/nodejs-poolController/discussions/605#discussioncomment-3646694.

**Reply by mojonba** [reply]
Hi, Can you please elaborate how can the export NODE_OPTIONS=--max_old_space_size=256 into the PM2 ecosystem.config.js file? I need it as well to run on a pi zero 2w. Thxs.

**Reply by tagyoureit** [reply]
```
module.exports = {
  apps: [
    {
      name: "my-app",
      script: "index.js",
      // limit Node.js memory to 256 MB
      env: {
        NODE_OPTIONS: "--max_old_space_size=256"
      }
    }
  ]
};
````
